### DISCUSSION TITLE
Masterpage "onpagestart" not called if table grows over one page

### DISCUSSION THREAD

**Discussion by migaman** [eclipse-birt/birt#1182] [Q&A] [answered]
Hello

My goal is to print a barcode on the footer of every page. The barcode contains the personal number of the current record.
I realized this with a Page Variable on the Master Page. The variable is set with the script "onPageStart". 
`reportContext.setPageVariable("NewVariableElement", row["persnr"]);`
A page break is made after each person. (Realized with Page break after group).

It works well except if the table data of the current record is larger than one page. In this case it seems the Script "onPageStart" is not called on the next Person. I view the report as a PDF.

Attached i made a pretty simple report with five csv data records. The "notes" detail in the table is only visible for personal number 1003.
[pagebreak.zip](https://github.com/eclipse/birt/files/10429748/pagebreak.zip)
When you view this as a PDF in the designer 4.10 (via Web Viewer or diretly "View Report as PDF") you can see that the footer matches with personal number until person 1003. 
The next page for person 1004 shows 1003. I would expect 1004.
The next next page for 1005 shows 1005. This is fine again.

I tried to set the page variable with other scripts like Table "onPageBreak" itself but it did not help.

Any ideas or shall I go for another approach to realize this?

**Comment by hvbtup** [comment]
Page variables only work correctly with separate `RunTask` and `RenderTask`, but not with `RunAndRenderTask`.

**Reply by migaman** [reply]
I'm testing it with Birt Report Designer RCP.
Isn't it `RunTask `and `RenderTask `when I press "View Report in Web Viewer" and after that Export as PDF in the Viewer?

When I do that it doesn't work for me. I made a new example with only three records.
[pagebreak.zip](https://github.com/eclipse/birt/files/10450539/pagebreak.zip)

Page 1 is OK: "1000" ist printed in the footer
Page 2 is OK: "1003" is printed in the footer
Page 3 is not OK: It is nothing printed in the footer. I expect 1003 because it is the second page of "1003" (same group)
Page 4 is OK: "1004" is printed in the footer.

What I figured out in the meantime: Setting the variable in the script for the group "onPageBreak" could solve my problem. 
`reportContext.setPageVariable("pers", "test");`

But I need to set a value from the data itself not a constant. Or the actual group value would also help me.
I already tried with `value`, `dataSetRow["persnr"]`, `row.persnr` with no success

**Comment by hvbtup** [comment]
There are obvious errors in your report.
First of all, `value` is not defined. What you probably mean is `row["persnr"]`.
Next, you should try to set the page variable in the `onCreate` event instead of in `onPageBreak`.

I don't think one can do anything useful in `onPageBreak`.

Disclaimer: I have hardly ever used page variables myself. Perhaps someone else can help you better.

**Comment by migaman** [comment]
OK thank you anyway. 

`value` works fine in my designer (4.10). `row["persnr"]` shows the next record. I tested this as well in 4.12 with the same result.
![Animation](https://user-images.githubusercontent.com/8789752/213437411-69d05cdf-f6d5-4aa4-935c-e5b8b3ff3061.gif)

`OnCreate` has no effect in my designer. The variable stays empty even if i set a constant like "test".

**Comment by speckyspooky** [comment, marked as answer]
Hello,

I checked your report and over all there comes different behavior elements of BIRT and the internal processing together.
Perhaps I have also a workaround for you which can help you.

At first some details to understand a little bit the special things.
The usage of the key-word "value" is valid and it is the shortcut of "this.value".
This works because it is located directly at the data element (used in "onPageBreak()").

Now the special processing things:
1. The page variables will be set to "null" on every new page and will "only" be set if the setting element included on the new page. The problem in your case is that the element "persnr" wasn't present at the "notes" page and so BIRT cleand your page variable and won't set  him because it is not located there.

2. The page generation is splitted of the content generation and both steps are not synchronized. The first step is to create the content because we have to know the full content and afterwards the pages will be generated.

3. Additional to it the problem with row["persnr"] is that the content process reads in every case the following record before the current page will be generated therefore record pointer of the dataset is set forward.


**Workaround**
I have added your "persnr" twice and set the specific visibility of all outputs to unvisibility.
So the element will be generated on the report side but all renderer will avoid the printing. But so you have filled the requirement of existing data element to set your page variable.
Please be aware dont' set the general visibility to unvisible. In this case the data element will be cutted from the generation directly and the workaround will not work.

Attached you will find my changed report version. My filed is the purple field.

![2023-01-21 21_47_30-_neu 4 - Notepad++](https://user-images.githubusercontent.com/41593722/213886623-b8faae65-57b8-4f25-8194-f82437742e45.png)


[new_report.zip](https://github.com/eclipse/birt/files/10473018/new_report.zip)

**Reply by migaman** [reply]
Hi 

Thanks for this solution and the detailed explanation. It is now clear for me and I could adapt this concept successfully to my original report
